lib/iris/_data_manager.py

```diff
@@ -0,0 +1,392 @@
+
+    def __copy__(self):
+        """
+        Forbid :class:`DataManager` instance shallow-copy support.
+
+        """
+        name = type(self).__name__
+        emsg = ('Shallow-copy of {!r} is not permitted. Use '
+                'copy.deepcopy() or {}.copy() instead.')
+        raise copy.Error(emsg.format(name, name))
+
+    def __deepcopy__(self, memo):
+        """
+        Allow :class:`DataManager` instance deepcopy support.
+
+        Args:
+
+        * memo:
+            :class:`copy` memo dictionary.
+
+        """
+        return self._deepcopy(memo)
+
+    def __repr__(self):
+        fmt = '{cls}({self.core_data!r}{dtype})'
+        dtype = ''
+        if self._realised_dtype is not None:
+            dtype = ', realised_dtype={!r}'.format(self._realised_dtype)
+        result = fmt.format(self=self, cls=type(self).__name__, dtype=dtype)
+        return result
+
+    def _assert_axioms(self):
+        """
+        Definition of the manager state, that should never be violated.
+
+        """
+        # Ensure there is a valid data state.
+        is_lazy = bool(self._lazy_array is not None)
+        is_real = bool(self._real_array is not None)
+        emsg = 'Unexpected data state, got {}lazy and {}real data.'
+        state = is_lazy ^ is_real
+        assert state, emsg.format('' if is_lazy else 'no ',
+                                  '' if is_real else 'no ')
+        # Ensure validity of realised dtype.
+        state = (self._realised_dtype is None or
+                 self._realised_dtype.kind in 'biu')
+        emsg = 'Unexpected realised dtype state, got {!r}'
+        assert state, emsg.format(self._realised_dtype)
+
+        # Ensure validity of lazy data with realised dtype.
+        state = not (self.has_real_data() and self._realised_dtype is not None)
+        emsg = ('Unexpected real data with realised dtype, got '
+                'real data and realised {!r}.')
+        assert state, emsg.format(self._realised_dtype)
+
+    def _deepcopy(self, memo, data=None, realised_dtype=None):
+        """
+        Perform a deepcopy of the :class:`DataManager` instance.
+
+        Args:
+
+        * memo:
+            :class:`copy` memo dictionary.
+
+        Kwargs:
+
+        * data:
+            Replacement data to substitute the currently managed
+            data with.
+
+        * realised_dtype:
+            Replacement for the intended dtype of the realised lazy data.
+
+        Returns:
+            :class:`DataManager`
+
+        """
+        try:
+            if data is None:
+                # Copy the managed data.
+                if self.has_lazy_data():
+                    data = copy.deepcopy(self._lazy_array, memo)
+                else:
+                    data = self._real_array.copy()
+            else:
+                # Check that the replacement data is valid relative to
+                # the currently managed data.
+                DataManager(self.core_data).replace(data)
```

Comment:
Sorry! This was actually a note to myself that I had meant to delete, before I submitted the review. I do get the purpose of this

>Actually, I stand-by the original, as it is a clean separation between checking the data replacement and realised_dtype replacement ... convince me otherwise :wink:

Any risk of memory problems? If you are working with big real data that you replace with more big data, you are creating an extra DataManager that is only a check but doesn't get used. Perhaps delete it?